```{r}
library(pacman)
p_load(data.table, dplyr, ggplot2, rtracklayer, RColorBrewer, scales, foreach, GenomicRanges,
viridis, annotatr, stringr, GGally, EnrichedHeatmap, circlize, corrplot, patchwork, splines)
options(scipen=999)
knitr::opts_chunk$set(fig.width=9, fig.height=7)
dir.create(outdatadir, recursive=TRUE)
source("./code/common.R")
```

# Profiles {.tabset}  

BlackListed regions are present/counted.  

```{r}
samples2do_BS <- c("WGBS_WT_met", "WGBS_hetero_met", "WGBS_homo_met", "WGBS_KO_met", "RRBS_hetero_met", "RRBS_homo_met")
samples2do_TOP <- c("WT", "Het", "Hom") 
coverages <- readRDS("./output/code/PrepAllData/mutantai_coverages2useTopData.RDS")
coverages_full <- coverages %>% 
  .[, end := start] %>% 
  .[chr %in% paste0("chr", c(1:19, "X", "Y"))] %>% 
  makeGRangesFromDataFrame(., keep.extra.column=TRUE)

coverages <- coverages %>% 
  .[WT>0, WT := 1] %>% 
  .[Het>0, Het := 1] %>% 
  .[Hom>0, Hom := 1] %>% 
  .[, end := start] %>% 
  .[, c("chr", "start", "end", samples2do_TOP), with=FALSE] %>% 
  .[chr %in% paste0("chr", c(1:19, "X", "Y"))] %>% 
  makeGRangesFromDataFrame(., keep.extra.column=TRUE)

tmp <- readRDS(paste0("output/code/PrepAllData//dAnnotations.RDS"))
repeatai <- tmp[["repeatai"]]
mcols(repeatai)$type <- paste0(mcols(repeatai)$Name, "_", mcols(repeatai)$Class, "_", mcols(repeatai)$Familly)
mcols(repeatai)$id <- paste0("element:", 1:length(repeatai))
annotations <- tmp[["anotacijos"]]
annotations <- annotations[seqnames(annotations) %in% paste0("chr", c(1:19, "X", "Y"))]
anotacijos <- c(annotations, repeatai)
CGI <- annotations[mcols(annotations)$type == "mm10_cpg_islands"]

tmp <- readRDS(paste0("output/code/PrepAllData//WGBS_vs_RRBS_data.RDS"))
met_gr <- tmp[["methylation_gr_th"]]
met_gr <- met_gr[seqnames(met_gr) %in% paste0("chr", c(1:19, "X", "Y"))]

rm(tmp)
gc()
```

```{r}
ext_target <- function(x){
  upai <- x[grepl("u", x$V1)]
  targetai <- x[grepl("t", x$V1)]
  targ_mod <- rbind(targetai, targetai) %>% .[, nr := gsub("t", "", V1) %>% as.numeric] %>% setkey(., nr) %>% .[, nr := NULL] %>% .[, V1 := paste0("t", 1:.N)] %>% .[]
  downai <- x[grepl("d", x$V1)]
  return(rbind(upai, targ_mod, downai))
}

get_profile <- function(BS_data, TOP_data, feat2do, vardas) {
  if(!(file.exists(paste0(outdatadir, vardas, "_profile.RDS")))) {
    extend <- 4000
    mat1 <- normalizeToMatrix(BS_data, feat2do, value_column = "WGBS_WT_met", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 100))
    mat2 <- normalizeToMatrix(BS_data, feat2do, value_column = "WGBS_hetero_met", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 100))
    mat3 <- normalizeToMatrix(BS_data, feat2do, value_column = "WGBS_homo_met", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 100))
    mat4 <- normalizeToMatrix(BS_data, feat2do, value_column = "WGBS_KO_met", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 100))
    mat5 <- normalizeToMatrix(BS_data, feat2do, value_column = "RRBS_hetero_met", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 100))
    mat6 <- normalizeToMatrix(BS_data, feat2do, value_column = "RRBS_homo_met", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 100))
    mat7 <- normalizeToMatrix(TOP_data, feat2do, value_column = "WT", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 1))
    mat8 <- normalizeToMatrix(TOP_data, feat2do, value_column = "Het", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 1))      
    mat9 <- normalizeToMatrix(TOP_data, feat2do, value_column = "Hom", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 1))                        
    saveRDS(list(mat1, mat2, mat3, mat4, mat5, mat6, mat7, mat8, mat9), paste0(outdatadir, vardas, "_profile.RDS"))
    } else {
    tmp <- readRDS(paste0(outdatadir, vardas, "_profile.RDS"))
    mat1 <- tmp[[1]]
    mat2 <- tmp[[2]]
    mat3 <- tmp[[3]]
    mat4 <- tmp[[4]]
    mat5 <- tmp[[5]]
    mat6 <- tmp[[6]]
    mat7 <- tmp[[7]]
    mat8 <- tmp[[8]]
    mat9 <- tmp[[9]]
    }
  row_tvarka <- as.data.table(colMeans(mat1[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target() %>% .[, V1]

  df <- rbind(
    as.data.table(colMeans(mat1[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target() %>% .[, sample := samples2do_BS[1]],
    as.data.table(colMeans(mat2[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := samples2do_BS[2]],
    as.data.table(colMeans(mat3[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target() %>% .[, sample := samples2do_BS[3]],
    as.data.table(colMeans(mat4[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := samples2do_BS[4]],
    as.data.table(colMeans(mat5[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := samples2do_BS[5]], 
    as.data.table(colMeans(mat6[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := samples2do_BS[6]], 
    as.data.table(colMeans(mat7[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := samples2do_TOP[1]] %>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat8[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target() %>% .[, sample := samples2do_TOP[2]] %>% .[, V2 := V2 * 100],
    as.data.table(colMeans(mat9[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target() %>% .[, sample := samples2do_TOP[3]] %>% .[, V2 := V2 * 100]
    ) %>% 
    as.data.table() %>% 
    .[, V1 := factor(V1, levels=row_tvarka)] %>% 
    .[, element := vardas]  
  return(df)
}

if(!file.exists(paste0(outdatadir, paste0("Profiles_matrix_elements.RDS")))) {
  temp <- rbind(
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_transcriptsCoding"], "transcripts"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_lncrna_gencode"], "lncRNA"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_processedPseudogene"], "procPseudoGene"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_UnProcessedPseudogene"], "unProcPseudoGene"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_enhancers_fantom"], "EnhancersFantom"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_ActiveEnhancers_Atlas"], "ActiveEnhancerAtlas"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_PoisedEnhancers_Atlas"], "PoisedEnhacerAtlas"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_ActiveEnhancers_fantom"], "ActiveEnhancerFantom"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_PoisedEnhancers_fantom"], "PoisedEnhacerFantom"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_enhancers"], "EnhancersAtlas"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_superEnhancer"], "superEnhacner"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_EPDpromoters"], "Promoters"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_EPDpromotersNC"], "PromotersNonCode"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_Silencer_E14"], "silencers_E14"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_Silencer_mESC"], "silencer_mESC"),
    get_profile(met_gr, coverages, annotations[annotations$type == "mm10_CTCF"], "CTCF")
  )
  saveRDS(temp, paste0(outdatadir, paste0("Profiles_matrix_elements.RDS")))
} else {temp <- readRDS(paste0(outdatadir, paste0("Profiles_matrix_elements.RDS")))}
  

if(!file.exists( paste0(outdatadir, paste0("Profiles_matrix_repeats.RDS")))) {
  set.seed(1234)
  LINEs <- repeatai[mcols(repeatai)$Class == "LINE"] %>% sample(5000, )
  set.seed(1234)
  SINEs <- repeatai[mcols(repeatai)$Class == "SINE"] %>% sample(5000, )
  set.seed(1234)
  LTR <- repeatai[mcols(repeatai)$Class == "LTR"] %>% sample(5000, )
  set.seed(1234)
  DNA <- repeatai[mcols(repeatai)$Class == "DNA"] %>% sample(5000, ) 
  set.seed(1234)
  Satellite  <- repeatai[mcols(repeatai)$Class == "Satellite"] %>% sample(5000, )
  set.seed(1234)
  Simple_repeat <- repeatai[mcols(repeatai)$Class == "Simple_repeat"] %>% sample(5000, )

  temp_reps<- rbind(
    get_profile(met_gr, coverages, LINEs, "LINE"),
    get_profile(met_gr, coverages, SINEs, "SINE"),
    get_profile(met_gr, coverages, LTR, "LTR"),
    get_profile(met_gr, coverages, DNA, "DNA"),
    get_profile(met_gr, coverages, Satellite, "Satellite"),
    get_profile(met_gr, coverages, Simple_repeat, "Simple_repeat")
  )
  saveRDS(temp_reps, paste0(outdatadir, paste0("Profiles_matrix_repeats.RDS"))) } else {
    temp_reps <- readRDS(paste0(outdatadir, paste0("Profiles_matrix_repeats.RDS")))
  }
```

## On elements

```{r, fig.width=8, fig.height=15}
temp %>% 
  .[, technology := "TopSeq"] %>% 
  .[grepl("RRBS", sample), technology := "RRBS"] %>% 
  .[grepl("WGBS", sample), technology := "WGBS"] %>% 
  .[, samplas := "Kita"] %>% 
  .[grepl("Het|het", sample),  samplas := "hetero"] %>% 
  .[grepl("Hom|homo", sample),  samplas := "homo"] %>%  
  .[grepl("WT", sample),  samplas := "WT"] %>%  
  .[sample != "WGBS_KO_met"] %>% 
  .[sample != "WGBS_WT_met"] %>% 
  .[!grepl("WT", sample), ] %>% 
ggplot(aes(V1, V2, group=samplas, colour=samplas)) +
      geom_line(size=0.5, alpha=0.75, lwd=1.25) +
      theme_Publication() +
      theme_bw() +
      ylab("Methylation/TopSeq*100") +
      theme(axis.title.x=element_blank(), 
            legend.position="bottom",
            panel.grid.major=element_blank(),
            panel.background=element_blank()
            ) +
      coord_cartesian(ylim=c(0, 100)) +
      scale_x_discrete(breaks = c("u100", "d1"), labels = c("FSS", "FES")) +
      scale_colour_manual(values = cols_samplesLong) +
      facet_grid(vars(element), vars(technology)) +
      geom_vline(xintercept=which(temp$V1 == 'u100'), linetype="dotted", colour="grey") +
      geom_vline(xintercept=which(temp$V1 == 'd1'), linetype="dotted", colour="grey") 
```

```{r, fig.width=8, fig.height=12}
temp_reps %>% 
  .[, technology := "TopSeq"] %>% 
  .[grepl("RRBS", sample), technology := "RRBS"] %>% 
  .[grepl("WGBS", sample), technology := "WGBS"] %>% 
  .[, samplas := "Kita"] %>% 
  .[grepl("Het|het", sample),  samplas := "hetero"] %>% 
  .[grepl("Hom|homo", sample),  samplas := "homo"] %>%  
  .[grepl("WT", sample),  samplas := "WT"] %>%  
  .[sample != "WGBS_KO_met"] %>% 
  .[sample != "WGBS_WT_met"] %>% 
  .[!grepl("WT", sample), ] %>% 
ggplot(aes(V1, V2, group=samplas, colour=samplas)) +
      geom_line(size=0.5, alpha=0.75, lwd=1.25) +
      theme_Publication() +
      theme_bw() +
      ylab("Methylation/TopSeq*100") +
      theme(axis.title.x=element_blank(), 
            legend.position="bottom",
            panel.grid.major=element_blank(),
            panel.background=element_blank()
            ) +
      coord_cartesian(ylim=c(0, 100)) +
      scale_x_discrete(breaks = c("u100", "d1"), labels = c("FSS", "FES")) +
      scale_colour_manual(values = cols_samplesLong) +
      facet_grid(vars(element), vars(technology)) +
      geom_vline(xintercept=which(temp$V1 == 'u100'), linetype="dotted", colour="grey") +
      geom_vline(xintercept=which(temp$V1 == 'd1'), linetype="dotted", colour="grey") 
```

## Specific CGI 

```{r}
p1 <- get_barPlot_bisulfite(makeGRangesFromDataFrame(data.frame(chr="chr11", start=3192997-1, end=3193699+1)), met_gr, "Sfi1", samples2do_BS, cols_samplesLong, samples2do_BS)
p2 <- get_barPlot_bisulfite(makeGRangesFromDataFrame(data.frame(chr="chr15", start=98256832-1, end=98257066+1)), met_gr, "H1fnt", samples2do_BS, cols_samplesLong, samples2do_BS)

(p1 | p2)  + plot_layout(guides = "collect", widths = c(2, 1)) & theme(legend.position = 'bottom')
```

Skirtumas tarp WT ir hetero poros bei atskirai homo ir KO. 

```{r}
p1 <- do_dif(makeGRangesFromDataFrame(data.frame(chr="chr11", start=3192997-1, end=3193699+1)), met_gr, "Sfi1. WT - hetero", "WGBS_WT_met", "WGBS_hetero_met")
p2 <- do_dif(makeGRangesFromDataFrame(data.frame(chr="chr11", start=3192997-1, end=3193699+1)), met_gr, "Sfi1. Homo - KO", "WGBS_homo_met", "WGBS_KO_met")
p3 <- do_dif(makeGRangesFromDataFrame(data.frame(chr="chr15", start=98256832-1, end=98257066+1)), met_gr, "H1fnt. WT - hetero", "WGBS_WT_met", "WGBS_hetero_met")
p4 <- do_dif(makeGRangesFromDataFrame(data.frame(chr="chr15", start=98256832-1, end=98257066+1)), met_gr, "H1fnt. Homo - KO", "WGBS_homo_met", "WGBS_KO_met")

(p1 | p2) / (p3 | p4)  + plot_layout(guides = "collect", widths = c(2, 1)) & theme(legend.position = 'bottom')

```


Y ašis skirtinga!!! (bet galima suvienodint)
```{r}
all_TOP <- c("Het_R2", "Het_R3", "Het_R4", "Hom_R1", "Hom_R2", "Hom_R3", "WT_R1", "WT_R2", "WT_R3")
p1 <- get_barplot_topSeq(makeGRangesFromDataFrame(data.frame(chr="chr11", start=3192997-1, end=3193699+1)), coverages_full, "Sfi1", all_TOP, cols_samples, samples2do_TOP)
p2 <- get_barplot_topSeq(makeGRangesFromDataFrame(data.frame(chr="chr15", start=98256832-1, end=98257066+1)), coverages_full, "H1fnt", all_TOP, cols_samples, samples2do_TOP)

(p1 | p2)   + plot_layout(guides = "collect", widths = c(2, 1)) & theme(legend.position = 'bottom')
```


## TOP-seq data 

Replication of Fig. 6A (). Dar daroma. 

```{r}
## repeatų:
get_per_Element <- function(elementas, TOP_data) {
  kiekis <- countOverlaps(elementas, TOP_data[mcols(TOP_data)$Het >= 0]) 
  s1 <- countOverlaps(elementas, TOP_data[mcols(TOP_data)$Het > 0])/kiekis 
  s2 <- countOverlaps(elementas, TOP_data[mcols(TOP_data)$Hom > 0])/kiekis 
  rez <- data.table(ID=mcols(elementas)$id, hetero=s1, homo=s2)
  nuliniai <- rez[hetero == 0, ID] 
  quant <- quantile(rez[hetero > 0, hetero] , na.rm=TRUE, c(0.0, 0.20, 0.40, 0.60, 0.80))

  return(list(nuliukai=elementas[mcols(elementas)$id %in% (rez[hetero == 0, ID])], 
              apatiniai=elementas[mcols(elementas)$id %in% (rez[(hetero > 0) & (hetero < quant[2]), ID])], 
              vidu=elementas[mcols(elementas)$id %in% (rez[(hetero >= quant[3]) & (hetero < quant[4]), ID])], 
              virsutiniai=elementas[mcols(elementas)$id %in% (rez[(hetero >= quant[5]), ID])]))
}

get_profile_groups <- function(grupes, TOP_data, vardas, dnmt3A, dnmt3B) {
  if(!(file.exists(paste0(outdatadir, vardas, "_GroupsProfile.RDS")))) {
    extend <- 4000
    mat1 <- normalizeToMatrix(TOP_data, grupes[[1]], value_column = "Het", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 1))
    mat2 <- normalizeToMatrix(TOP_data, grupes[[1]], value_column = "Hom", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 1))
    mat3 <- normalizeToMatrix(TOP_data, grupes[[2]], value_column = "Het", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 1))
    mat4 <- normalizeToMatrix(TOP_data, grupes[[2]], value_column = "Hom", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 1))            
    mat5 <- normalizeToMatrix(TOP_data, grupes[[3]], value_column = "Het", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 1))
    mat6 <- normalizeToMatrix(TOP_data, grupes[[3]], value_column = "Hom", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 1))
    mat7 <- normalizeToMatrix(TOP_data, grupes[[4]], value_column = "Het", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 1))
    mat8 <- normalizeToMatrix(TOP_data, grupes[[4]], value_column = "Hom", mean_mode = "absolute",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 1))  
    mat9 <- normalizeToMatrix(dnmt3A, grupes[[1]], value_column = "score", mean_mode = "coverage",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 10)) 
    mat10 <- normalizeToMatrix(dnmt3A, grupes[[2]], value_column = "score", mean_mode = "coverage",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 10))             
    mat11 <- normalizeToMatrix(dnmt3A, grupes[[3]], value_column = "score", mean_mode = "coverage",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 10))                         
    mat12 <- normalizeToMatrix(dnmt3A, grupes[[4]], value_column = "score", mean_mode = "coverage",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 10))                         
    mat13 <- normalizeToMatrix(dnmt3B, grupes[[1]], value_column = "score", mean_mode = "coverage",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 10)) 
    mat14 <- normalizeToMatrix(dnmt3B, grupes[[2]], value_column = "score", mean_mode = "coverage",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 10))             
    mat15 <- normalizeToMatrix(dnmt3B, grupes[[3]], value_column = "score", mean_mode = "coverage",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 10))                         
    mat16 <- normalizeToMatrix(dnmt3B, grupes[[4]], value_column = "score", mean_mode = "coverage",
            extend = extend, background = NA, smooth = FALSE, target_ratio = 0.3, limit=c(0, 10))                         
    saveRDS(list(mat1, mat2, mat3, mat4, mat5, mat6, mat7, mat8, mat9, mat10, mat11, mat12, mat13, mat14, mat15, mat16), paste0(outdatadir, vardas, "_GroupsProfile.RDS"))
    } else {
    tmp <- readRDS(paste0(outdatadir, vardas, "_GroupsProfile.RDS"))
    mat1 <- tmp[[1]]
    mat2 <- tmp[[2]]
    mat3 <- tmp[[3]]
    mat4 <- tmp[[4]]
    mat5 <- tmp[[5]]
    mat6 <- tmp[[6]]
    mat7 <- tmp[[7]]
    mat8 <- tmp[[8]]
    mat9 <- tmp[[9]]
    mat10 <- tmp[[10]]
    mat11 <- tmp[[11]]
    mat12 <- tmp[[12]]
    mat13 <- tmp[[13]]
    mat14 <- tmp[[14]]
    mat15 <- tmp[[15]]
    mat16 <- tmp[[16]]
    }
  row_tvarka <- as.data.table(colMeans(mat1[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target() %>% .[, V1]

  df <- rbind(
    as.data.table(colMeans(mat1[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target() %>% .[, sample := "Het"] %>% .[, grupe := "0%"] %>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat2[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := "Hom"] %>% .[, grupe := "0%"] %>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat3[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target() %>% .[, sample := "Het"] %>% .[, grupe := "20%"] %>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat4[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := "Hom"] %>% .[, grupe := "20%"] %>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat5[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := "Het"] %>% .[, grupe := "40-60%"] %>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat6[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := "Hom"] %>% .[, grupe := "40-60%"] %>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat7[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := "Het"] %>% .[, grupe := "80%"]%>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat8[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target() %>% .[, sample := "Hom"] %>% .[, grupe := "80%"]%>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat9[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target() %>% .[, sample := "dnmt3A"] %>% .[, grupe := "0%"] %>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat13[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := "dnmt3B"] %>% .[, grupe := "0%"] %>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat10[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target() %>% .[, sample := "dnmt3A"] %>% .[, grupe := "20%"] %>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat14[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := "dnmt3B"] %>% .[, grupe := "20%"] %>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat11[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := "dnmt3A"] %>% .[, grupe := "40-60%"] %>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat15[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := "dnmt3B"] %>% .[, grupe := "40-60%"] %>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat12[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target()%>% .[, sample := "dnmt3A"] %>% .[, grupe := "80%"]%>% .[, V2 := V2 * 100], 
    as.data.table(colMeans(mat16[,], na.rm=TRUE), keep.rownames=TRUE) %>% ext_target() %>% .[, sample := "dnmt3B"] %>% .[, grupe := "80%"]%>% .[, V2 := V2 * 100]
    ) %>% 
    as.data.table() %>% 
    .[, V1 := factor(V1, levels=row_tvarka)] %>% 
    .[, element := vardas]  
  return(df)
}

dnmt3A <- import("input/OtherData/GSM3772691_mESC_Par_DNMT3A.bw")
dnmt3B <- import("input/OtherData/GSM3772694_mESC_Par_DNMT3B.bw")

if(!file.exists((paste0(outdatadir, "Profiles_regu_groups.RDS")))) {
temp <- rbind(
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_EPDpromoters"],  coverages), coverages, 'EPDpromoter', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_Silencer_E14"],  coverages), coverages, 'Silencers_E14', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_Silencer_mESC"],  coverages), coverages, 'Silencers_mESC', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_enhancers_fantom"],  coverages), coverages, 'enhancers_fantom', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_enhancers"],  coverages), coverages, 'enhancers_atlas', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(CGI[countOverlaps(CGI, annotations[mcols(annotations)$type == "mm10_EPDpromoters"]) > 0, ],  coverages), coverages, 'CGI promoters', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(CGI[(countOverlaps(CGI, annotations[mcols(annotations)$type == "mm10_transcriptsCoding"]) == 0) & (countOverlaps(CGI, annotations[mcols(annotations)$type == "mm10_EPDpromoters"]) == 0), ],  coverages), coverages, 'CGI intergenic', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(CGI[countOverlaps(CGI, annotations[mcols(annotations)$type == "mm10_transcriptsCoding"]) > 0, ]  ,  coverages), coverages, 'CGI intragenics', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_CTCF"],  coverages), coverages, 'mm10_CTCF', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_proteinCoding"],  coverages), coverages, 'proteinCoding', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_lncrna_gencode"],  coverages), coverages, 'lncrna_gencode', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_lncRNA_vM23"],  coverages), coverages, 'lncRNA_vM23', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_UnProcessedPseudogene"],  coverages), coverages, 'UnProcessedPseudogene', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_processedPseudogene"],  coverages), coverages, 'processedPseudogene', dnmt3A, dnmt3B)

)
saveRDS(temp, paste0(outdatadir, "Profiles_regu_groups.RDS")) } else {temp <- readRDS(paste0(outdatadir, "Profiles_regu_groups.RDS"))}

if(!(file.exists(paste0(outdatadir, "Profiles_repeatai_groups.RDS")))) {
  tempR <- rbind(
    get_profile_groups(get_per_Element(repeatai[mcols(repeatai)$Class == "LINE"],  coverages), coverages, 'LINE', dnmt3A, dnmt3B),
    get_profile_groups(get_per_Element(repeatai[mcols(repeatai)$Class == "SINE"],  coverages), coverages, 'SINE', dnmt3A, dnmt3B),
    get_profile_groups(get_per_Element(repeatai[mcols(repeatai)$Class == "LTR"],  coverages), coverages, 'LTR', dnmt3A, dnmt3B),
    get_profile_groups(get_per_Element(repeatai[mcols(repeatai)$Class == "Satellite"],  coverages), coverages, 'Satellite', dnmt3A, dnmt3B),
    get_profile_groups(get_per_Element(repeatai[mcols(repeatai)$Class == "DNA"],  coverages), coverages, 'DNA', dnmt3A, dnmt3B),
    get_profile_groups(get_per_Element(repeatai[mcols(repeatai)$Class == "Simple_repeat"],  coverages), coverages, 'Simple_repeat', dnmt3A, dnmt3B)
  )
  saveRDS(tempR, paste0(outdatadir, "Profiles_repeatai_groups.RDS")) } else {tempR <- readRDS(paste0(outdatadir, "Profiles_repeatai_groups.RDS"))}

```

```{r, fig.width=12, fig.height=8}
temp %>% 
  .[sample %in% c("Het", "Hom"), ] %>%  
  ggplot(aes(V1, V2, group=grupe, colour=grupe)) +
      geom_line(size=0.5, alpha=0.75, lwd=1.25) +
      theme_Publication() +
      theme_bw() +
      ylab("Identified CpG, %") +
      theme(axis.title.x=element_blank(), 
            legend.position="bottom",
            panel.grid.major=element_blank(),
            panel.background=element_blank()
            ) +
      coord_cartesian(ylim=c(0, 100)) +
      scale_x_discrete(breaks = c("u100", "d1"), labels = c("FSS", "FES")) +
#      scale_colour_manual(values = cols_samplesLong) +
      facet_grid(vars(sample), vars(element)) +
      geom_vline(xintercept=which(temp$V1 == 'u100'), linetype="dotted", colour="grey") +
      geom_vline(xintercept=which(temp$V1 == 'd1'), linetype="dotted", colour="grey") 
```

```{r, fig.width=12, fig.height=8}
tempR %>% 
  .[sample %in% c("Het", "Hom"), ] %>%  
  ggplot(aes(V1, V2, group=grupe, colour=grupe)) +
      geom_line(size=0.5, alpha=0.75, lwd=1.25) +
      theme_Publication() +
      theme_bw() +
      ylab("Identified CpG, %") +
      theme(axis.title.x=element_blank(), 
            legend.position="bottom",
            panel.grid.major=element_blank(),
            panel.background=element_blank()
            ) +
      coord_cartesian(ylim=c(0, 100)) +
      scale_x_discrete(breaks = c("u100", "d1"), labels = c("FSS", "FES")) +
#      scale_colour_manual(values = cols_samplesLong) +
      facet_grid(vars(sample), vars(element)) +
      geom_vline(xintercept=which(temp$V1 == 'u100'), linetype="dotted", colour="grey") +
      geom_vline(xintercept=which(temp$V1 == 'd1'), linetype="dotted", colour="grey") 
```


## ChIP-seq data 

Duomenys is GEO. 

```{r, fig.width=12, fig.height=8}
temp %>% 
  .[!(sample %in% c("Het", "Hom")), ] %>%  
  ggplot(aes(V1, V2, group=grupe, colour=grupe)) +
      geom_line(size=0.5, alpha=0.75, lwd=1.25) +
      theme_Publication() +
      theme_bw() +
      ylab("Coverage, reads") +
      theme(axis.title.x=element_blank(), 
            legend.position="bottom",
            panel.grid.major=element_blank(),
            panel.background=element_blank()
            ) +
      coord_cartesian(ylim=c(200, 800)) +
      scale_x_discrete(breaks = c("u100", "d1"), labels = c("FSS", "FES")) +
#      scale_colour_manual(values = cols_samplesLong) +
      facet_grid(vars(sample), vars(element)) +
      geom_vline(xintercept=which(temp$V1 == 'u100'), linetype="dotted", colour="grey") +
      geom_vline(xintercept=which(temp$V1 == 'd1'), linetype="dotted", colour="grey") 

```


```{r, fig.width=12, fig.height=8}
tempR %>% 
  .[!(sample %in% c("Het", "Hom")), ] %>%  
  ggplot(aes(V1, V2, group=grupe, colour=grupe)) +
      geom_line(size=0.5, alpha=0.75, lwd=1.25) +
      theme_Publication() +
      theme_bw() +
      ylab("Coverage, reads") +
      theme(axis.title.x=element_blank(), 
            legend.position="bottom",
            panel.grid.major=element_blank(),
            panel.background=element_blank()
            ) +
      coord_cartesian(ylim=c(200, 800)) +
      scale_x_discrete(breaks = c("u100", "d1"), labels = c("FSS", "FES")) +
#      scale_colour_manual(values = cols_samplesLong) +
      facet_grid(vars(sample), vars(element)) +
      geom_vline(xintercept=which(temp$V1 == 'u100'), linetype="dotted", colour="grey") +
      geom_vline(xintercept=which(temp$V1 == 'd1'), linetype="dotted", colour="grey") 
```

## ChIP-Seq II

Principas kaip ir ChIP-seq, bet čia duomenys mano pačios sumappinti ir sunormalizuoti. 

```{r}
dnmt3A <- import("input/OtherData/RAW_ChIP/SRR9074481.bw")
dnmt3B <- import("input/OtherData/RAW_ChIP/SRR9074484.bw")

if(!file.exists((paste0(outdatadir, "Profiles_regu_groupsMano.RDS")))) {
temp <- rbind(
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_EPDpromoters"],  coverages), coverages, 'EPDpromoterMano', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_Silencer_E14"],  coverages), coverages, 'Silencers_E14Mano', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_Silencer_mESC"],  coverages), coverages, 'Silencers_mESCMano', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_enhancers_fantom"],  coverages), coverages, 'enhancers_fantomMano', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_enhancers"],  coverages), coverages, 'enhancers_atlasMano', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(CGI[countOverlaps(CGI, annotations[mcols(annotations)$type == "mm10_EPDpromoters"]) > 0, ],  coverages), coverages, 'CGI promotersMano', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(CGI[(countOverlaps(CGI, annotations[mcols(annotations)$type == "mm10_transcriptsCoding"]) == 0) & (countOverlaps(CGI, annotations[mcols(annotations)$type == "mm10_EPDpromoters"]) == 0), ],  coverages), coverages, 'CGIintergenicMano', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(CGI[countOverlaps(CGI, annotations[mcols(annotations)$type == "mm10_transcriptsCoding"]) > 0, ]  ,  coverages), coverages, 'CGIintragenicsMano', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_CTCF"],  coverages), coverages, 'mm10_CTCFMano', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_proteinCoding"],  coverages), coverages, 'proteinCodingMano', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_lncrna_gencode"],  coverages), coverages, 'lncrna_gencodeMano', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_lncRNA_vM23"],  coverages), coverages, 'lncRNA_vM23Mano', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_UnProcessedPseudogene"],  coverages), coverages, 'UnProcessedPseudogeneMano', dnmt3A, dnmt3B),
  get_profile_groups(get_per_Element(annotations[mcols(annotations)$type == "mm10_processedPseudogene"],  coverages), coverages, 'processedPseudogeneMano', dnmt3A, dnmt3B)

)
saveRDS(temp, paste0(outdatadir, "Profiles_regu_groupsMano.RDS")) } else {temp <- readRDS(paste0(outdatadir, "Profiles_regu_groupsMano.RDS"))}

if(!(file.exists(paste0(outdatadir, "Profiles_repeatai_groupsMano.RDS")))) {
  tempR <- rbind(
    get_profile_groups(get_per_Element(repeatai[mcols(repeatai)$Class == "LINE"],  coverages), coverages, 'LINEMano', dnmt3A, dnmt3B),
    get_profile_groups(get_per_Element(repeatai[mcols(repeatai)$Class == "SINE"],  coverages), coverages, 'SINEMano', dnmt3A, dnmt3B),
    get_profile_groups(get_per_Element(repeatai[mcols(repeatai)$Class == "LTR"],  coverages), coverages, 'LTRMano', dnmt3A, dnmt3B),
    get_profile_groups(get_per_Element(repeatai[mcols(repeatai)$Class == "Satellite"],  coverages), coverages, 'SatelliteMano', dnmt3A, dnmt3B),
    get_profile_groups(get_per_Element(repeatai[mcols(repeatai)$Class == "DNA"],  coverages), coverages, 'DNAMano', dnmt3A, dnmt3B),
    get_profile_groups(get_per_Element(repeatai[mcols(repeatai)$Class == "Simple_repeat"],  coverages), coverages, 'Simple_repeatMano', dnmt3A, dnmt3B)
  )
  saveRDS(tempR, paste0(outdatadir, "Profiles_repeatai_groupsMano.RDS")) } else {tempR <- readRDS(paste0(outdatadir, "Profiles_repeatai_groupsMano.RDS"))}
```


```{r, fig.width=12, fig.height=8}
temp %>% 
  .[!(sample %in% c("Het", "Hom")), ] %>%  
  ggplot(aes(V1, V2, group=grupe, colour=grupe)) +
      geom_line(size=0.5, alpha=0.75, lwd=1.25) +
      theme_Publication() +
      theme_bw() +
      ylab("Coverage, reads") +
      theme(axis.title.x=element_blank(), 
            legend.position="bottom",
            panel.grid.major=element_blank(),
            panel.background=element_blank()
            ) +
      #coord_cartesian(ylim=c(200, 800)) +
      scale_x_discrete(breaks = c("u100", "d1"), labels = c("FSS", "FES")) +
#      scale_colour_manual(values = cols_samplesLong) +
      facet_grid(vars(sample), vars(element)) +
      geom_vline(xintercept=which(temp$V1 == 'u100'), linetype="dotted", colour="grey") +
      geom_vline(xintercept=which(temp$V1 == 'd1'), linetype="dotted", colour="grey") 

```


```{r, fig.width=12, fig.height=8}
tempR %>% 
  .[!(sample %in% c("Het", "Hom")), ] %>%  
  ggplot(aes(V1, V2, group=grupe, colour=grupe)) +
      geom_line(size=0.5, alpha=0.75, lwd=1.25) +
      theme_Publication() +
      theme_bw() +
      ylab("Coverage, reads") +
      theme(axis.title.x=element_blank(), 
            legend.position="bottom",
            panel.grid.major=element_blank(),
            panel.background=element_blank()
            ) +
   #   coord_cartesian(ylim=c(200, 800)) +
      scale_x_discrete(breaks = c("u100", "d1"), labels = c("FSS", "FES")) +
#      scale_colour_manual(values = cols_samplesLong) +
      facet_grid(vars(sample), vars(element)) +
      geom_vline(xintercept=which(temp$V1 == 'u100'), linetype="dotted", colour="grey") +
      geom_vline(xintercept=which(temp$V1 == 'd1'), linetype="dotted", colour="grey") 
```